search/feature_offset_match.hpp

```diff
@@ -125,95 +49,82 @@ bool FindLangIndex(trie::Iterator<ValueList<TValue>> const & trieRoot, uint8_t l
 }
 }  // namespace
 
-template <typename TValue, typename TF>
-void FullMatchInTrie(trie::Iterator<ValueList<TValue>> const & trieRoot,
-                     strings::UniChar const * rootPrefix, size_t rootPrefixSize,
-                     strings::UniString s, TF & f)
+template <typename Value, typename DFAIt, typename ToDo>
+bool MatchInTrie(trie::Iterator<ValueList<Value>> const & trieRoot,
+                 strings::UniChar const * rootPrefix, size_t rootPrefixSize, DFAIt dfa,
+                 ToDo && toDo)
 {
-  if (!CheckMatchString(rootPrefix, rootPrefixSize, s, false /* prefix */))
-    return;
+  using TrieDFA = shared_ptr<trie::Iterator<ValueList<Value>>>;
+  using State = pair<TrieDFA, DFAIt>;
 
-  size_t symbolsMatched = 0;
-  bool bFullEdgeMatched;
-  auto const it = MoveTrieIteratorToString(trieRoot, s, symbolsMatched, bFullEdgeMatched);
-
-  if (!it || (!s.empty() && !bFullEdgeMatched) || symbolsMatched != s.size())
-    return;
+  for (size_t i = 0; i < rootPrefixSize; ++i)
+    dfa.Move(rootPrefix[i]);
 
-#if defined(OMIM_OS_IPHONE) && !defined(__clang__)
-  // Here is the dummy mutex to avoid mysterious iOS GCC-LLVM bug here.
-  static threads::Mutex dummyM;
-  threads::MutexGuard dummyG(dummyM);
-#endif
-
-  ASSERT_EQUAL(symbolsMatched, s.size(), ());
-
-  it->m_valueList.ForEach(f);
-}
+  if (dfa.Rejects())
+    return false;
 
-template <typename TValue, typename TF>
-void PrefixMatchInTrie(trie::Iterator<ValueList<TValue>> const & trieRoot,
-                       strings::UniChar const * rootPrefix, size_t rootPrefixSize,
-                       strings::UniString s, TF & f)
-{
-  if (!CheckMatchString(rootPrefix, rootPrefixSize, s, true /* prefix */))
-    return;
+  queue<State> q;
+  q.emplace(trieRoot.Clone(), dfa);
 
-  using TIterator = trie::Iterator<ValueList<TValue>>;
+  bool found = false;
 
-  using TQueue = vector<shared_ptr<TIterator>>;
-  TQueue trieQueue;
+  while (!q.empty())
   {
-    size_t symbolsMatched = 0;
-    bool bFullEdgeMatched;
-    auto const it = MoveTrieIteratorToString(trieRoot, s, symbolsMatched, bFullEdgeMatched);
-
-    UNUSED_VALUE(symbolsMatched);
-    UNUSED_VALUE(bFullEdgeMatched);
+    auto const p = q.front();
+    q.pop();
 
-    if (!it)
-      return;
+    auto const & trie = p.first;
+    auto const & dfa = p.second;
 
-    trieQueue.push_back(it);
-  }
+    if (dfa.Accepts())
+    {
+      trie->m_valueList.ForEach(toDo);
+      found = true;
+    }
 
-  while (!trieQueue.empty())
-  {
-    auto const it = trieQueue.back();
-    trieQueue.pop_back();
+    size_t const numEdges = trie->m_edge.size();
+    for (size_t i = 0; i < numEdges; ++i)
+    {
+      auto const & edge = trie->m_edge[i];
 
-    it->m_valueList.ForEach(f);
+      auto cur = dfa;
```

Comment:
Why? What's wrong? We have current position, edge, and moved current position by edge. From my point of view, temp is not good and not readable, copy is even worse.